lib/elixir/lib/module/parallel_checker.ex

```diff
@@ -0,0 +1,268 @@
+  `:error` if the MFA is not deprecated.
+  """
+  @spec fetch_deprecated(cache(), module(), atom(), arity()) :: {:ok, binary()} | :error
+  def fetch_deprecated({_server, ets}, module, fun, arity) do
+    # This is only called after we have checked undefined
+    # so we can assume the MFA exists
+    case :ets.lookup(ets, {:export, {module, fun, arity}}) do
+      [{_key, _kind, nil}] -> :error
+      [{_key, _kind, reason}] -> {:ok, reason}
+    end
+  end
+
+  @doc """
+  Returns all exported functions and macros for the given module from
+  the cache.
+  """
+  @spec all_exports(cache(), module()) :: [{atom(), arity()}]
+  def all_exports({_server, ets}, module) do
+    # This is only called after we get a deprecation notice
+    # so we can assume it's a cached module
+    [{_key, exports}] = :ets.lookup(ets, {:all_exports, module})
+
+    [{{:__info__, 1}, :def} | exports]
+    |> Enum.map(fn {function, _kind} -> function end)
+    |> Enum.sort()
+  end
+
+  @doc """
+  Collects all exported functions and macros from the module definition ASTs.
+  """
+  @spec definitions_to_exports([{atom(), arity(), term(), term()}]) ::
+          [{{atom(), arity()}, kind()}]
+  def definitions_to_exports(definitions) do
+    Enum.flat_map(definitions, fn {function, kind, _meta, _clauses} ->
+      if kind in [:def, :defmacro] do
+        [{function, kind}]
+      else
+        []
+      end
+    end)
+  end
+
+  def init([modules, schedulers]) do
+    ets = :ets.new(:checker_cache, [:set, :public, {:read_concurrency, true}])
+    preload_cache(ets, modules)
+
+    state = %{
+      ets: ets,
+      waiting: %{},
+      reply_when_finished: nil,
+      modules: modules,
+      spawned: 0,
+      schedulers: schedulers,
+      binaries: [],
+      warnings: []
+    }
+
+    {:ok, state}
+  end
+
+  def handle_call(:start_and_wait, from, %{reply_when_finished: nil} = state) do
+    state = spawn_checkers(%{state | reply_when_finished: from})
+    {:noreply, state}
+  end
+
+  def handle_call({:lock, module}, from, %{waiting: waiting} = state) do
+    case waiting do
+      %{^module => froms} ->
+        waiting = :maps.put(module, [from | froms], state.waiting)
+        {:noreply, %{state | waiting: waiting}}
+
+      %{} ->
+        waiting = :maps.put(module, [], state.waiting)
+        {:reply, true, %{state | waiting: waiting}}
+    end
+  end
+
+  def handle_call({:unlock, module}, _from, %{waiting: waiting} = state) do
+    froms = :maps.get(module, waiting)
+    Enum.each(froms, &:gen_server.reply(&1, false))
+    waiting = :maps.remove(module, waiting)
+    {:reply, :ok, %{state | waiting: waiting}}
+  end
+
+  def handle_info({__MODULE__, module, binary, warnings}, state) do
+    state = %{
+      state
+      | binaries: [{module, binary} | state.binaries],
+        warnings: warnings ++ state.warnings,
+        spawned: state.spawned - 1
+    }
+
+    if state.spawned == 0 and state.modules == [] and !!state.reply_when_finished do
```

Comment:
```suggestion
    if state.spawned == 0 and state.modules == [] and not state.reply_when_finished do
```

Reply:
This does not work since `reply_when_finished` is of type `GenServer.from() | nil`.